cleveldb_batch.go

```diff
@@ -10,33 +10,60 @@ type cLevelDBBatch struct {
 	batch *levigo.WriteBatch
 }
 
+func newCLevelDBBatch(db *CLevelDB) *cLevelDBBatch {
+	return &cLevelDBBatch{
+		db:    db,
+		batch: levigo.NewWriteBatch(),
+	}
+}
+
 // Set implements Batch.
-func (b *cLevelDBBatch) Set(key, value []byte) {
+func (b *cLevelDBBatch) Set(key, value []byte) error {
+	if b.batch == nil {
+		return ErrBatchClosed
+	}
 	b.batch.Put(key, value)
+	return nil
 }
 
 // Delete implements Batch.
-func (b *cLevelDBBatch) Delete(key []byte) {
+func (b *cLevelDBBatch) Delete(key []byte) error {
+	if b.batch == nil {
+		return ErrBatchClosed
+	}
 	b.batch.Delete(key)
+	return nil
 }
 
 // Write implements Batch.
 func (b *cLevelDBBatch) Write() error {
-	if err := b.db.db.Write(b.db.wo, b.batch); err != nil {
+	if b.batch == nil {
+		return ErrBatchClosed
+	}
+	err := b.db.db.Write(b.db.wo, b.batch)
+	if err != nil {
 		return err
 	}
-	return nil
+	return b.Close()
```

Comment:
do we even need a separate Close method? if we execute Close in Write/WriteSync

Reply:
Yeah, because the caller may e.g. return an error while they are building the batch, in which case they need to close it without calling `Write`.

Reply:
I can easily envision cases where Close will be called twice... maybe it's okay, but I'd prefer to remove Close calls from Write/WriteSync. You're mixing responsibilities

Reply:
Calling `Close()` twice is fine, it is idempotent. It is a normal Go pattern to use `defer c.Close()` and then also do an explicit `c.Close()` as early as possible to free up resources before the function exits.

If we don't call `Close()` in `Write()`, then what should happen when a caller calls `Write()` twice?

Reply:
> It is a normal Go pattern to use defer c.Close() for error handling and then also do an explicit c.Close() as early as possible to free up resources before the function exits.

Didn't know about such.

> If we don't call Close() in Write(), then what should happen when a caller calls Write() twice?

you can't prevent people from shooting themselves into the leg. I'd assume correct usage in most cases.

New -> Set/Delete -> Write(Sync) -> Close (via defer)

Reply:
> you can't prevent people from shooting themselves into the leg. I'd assume correct usage in most cases.

Of course, we just have to specify what should happen, as unspecified behavior is the root of all evil. Either `Write()` also calls `Close()`, in which case a second `Write()` call errors, otherwise the second `Write()`call should write the same batch again.

I think erroring is safer, and easier to implement across all backends. The typical and recommended usage would still be the same flow that you outline.

Reply:
> I think erroring is safer, and easier to implement across all backends. 

Agree.

Reply:
Ok, I think this should be good to go then.